### DISCUSSION TITLE
INVALID_IMAGE_OPTIMIZE_REQUEST on Vercel Deployment: While using an internal nextjs api route with slug as src

### DISCUSSION THREAD

**Discussion by 0xxmemo** [vercel/next.js#70043] [Help]
### Link to the code that reproduces this issue

https://github.com/mguleryuz/next_api-img_opt-issue-rep

### To Reproduce

1. Start the application using `next dev` | `next build && next start` the image loads on localhost
2. Navigate to [Vercel Deployed App](https://next-api-img-opt-issue-rep.vercel.app) you will see that the image is not loading

### Current vs. Expected behavior

I expect that the optimized image loads on Vercel Deployment as it does on localhost prod or dev

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:04 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T8122
  Available memory (MB): 16384
  Available CPU cores: 8
Binaries:
  Node: 20.17.0
  npm: 10.8.2
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.2.10 // Latest available version is detected (14.2.10).
  eslint-config-next: 14.2.10
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.6.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Image (next/image)

### Which stage(s) are affected? (Select all that apply)

Vercel (Deployed)

**Comment by samcx** [comment]
@mguleryuz I've seen this error before (during my time as a Customer Success Engineer)—it's something with your image or upstream provider, not with Next.js.

I recommend going through `vercel.com/help` (the AI might already know the answer) or https://vercel.community if you need more help with this!

**Reply by 0xxmemo** [reply]
Hello @samcx I have tried multiple type of images, and I have researched the issue at every corner of the web for the last 2 days.

The only way I was able to solve this on Vercel Production was to pass the Host Origin down to all my components and add the origin suffix to the image src's.

**Reply by 0xxmemo** [reply]
This problem is mainly to do with the [slug] without the slug the image loads on Vercel Deployment as well, I will update the deployment to demonstrate that as well.

**Reply by samcx** [reply]
Your static image `vercel.png` should be inside a `/public` directory → https://nextjs.org/docs/app/building-your-application/optimizing/static-assets.

Is there a reason why you're trying to read a file from a Route Handler?

The image will already be cached forever when inside `/public`, since it's a static asset. Also, in the filesystem when deployed, that image will not exist. A Route Handler, which ends up getting built as a Vercel Function on Vercel, does not have the same access to that file (not the same filesystem) as your local directory.

**Reply by 0xxmemo** [reply]
This was only for demonstration purposes my actual setup is using a MongoDB bucket to retrieve the images.

For the second point you made it does work without the [slug] so I do not think it is to do with file access.

**Reply by 0xxmemo** [reply]
@samcx I have updated the repo and the Vercel deployment so you can see the different behavior between the slug and non slug api routes in the context of next.js image optimization.

I still think this is a real issue and has to be looked into, please after making sure that my investigations are correct move the issue out of discussions and back into the issues section I don't think I have spent 2 days for nothing 😂.

**Reply by samcx** [reply]
> This was only for demonstration purposes my actual setup is using a MongoDB bucket to retrieve the images.

Did you also make sure to add a `remotePattern` to your `next.config` → https://nextjs.org/docs/app/building-your-application/optimizing/images#remote-images? In your case, I don't think you need to use a Route Handler to retrieve your images.

> For the second point you made it does work without the [slug]

I think this may only be working because the Route Handler is static by default (true for v14 and below) and cached at **build** time. Whereas if you use `params`, the Route Handler is now dynamic, and no longer  is in the same filesystem at **request** time.

Nonetheless, it's not the recommended path.

**Reply by 0xxmemo** [reply]
> Did you also make sure to add a remotePattern to your next.config → https://nextjs.org/docs/app/building-your-application/optimizing/images#remote-images? In your case, I don't think you need to use a Route Handler to retrieve your images.

Yes I have tried with wild card and non wild card ones, im updating the case repo to also include those.

> I think this may only be working because the Route Handler is static by default (true for v14 and below) and cached at build time. Nonetheless, it's not the recommended path.

I will also update the routes to fetch the image from some other source to showcase how its not about the file being buffered down

**Reply by 0xxmemo** [reply]
@samcx the repo is refactored to fetch the png instead of use the local png file, also next.config is refactored to include **.vercel.app hostname in `remotePatterns`.

**Reply by samcx** [reply]
I would also refactor to not use a Route Handler. A Route Handler that's static essentially runs in a different environment than a dynamic one.

This is the error message we get internally (on Vercel).

```
ImageError with status 401, url "https://next-api-img-opt-issue-<blurring hash>.vercel.app/api/image/gstatic", and message "Bad status code from upstream image"
```

Your Image Provider is likely blocking the IP when a Vercel Function (the [slug] case) is calling that image, but allowing the request in the non-[slug] case.
